src/bsoncxx/private/error_category.cpp

```diff
@@ -0,0 +1,153 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <bsoncxx/config/prelude.hpp>
+
+#include <system_error>
+
+#include <bson.h>
+
+#include <bsoncxx/private/error_category.hpp>
+
+#include <bsoncxx/exception/error_code.hpp>
+#include <bsoncxx/private/libbson_error.hpp>
+
+namespace {
+
+using namespace bsoncxx;
+
+///
+/// An error category for errors originating in bsoncxx.
+///
+class bsoncxx_error_category_impl final : public std::error_category {
+   public:
+    const char* name() const noexcept override {
+        return "bsoncxx";
+    }
+
+    std::string message(int condition) const noexcept override {
+        switch (static_cast<error_code>(condition)) {
+            case error_code::k_cannot_append_key_in_sub_array:
+                return "tried to append new key while in subarray";
+            case error_code::k_cannot_close_array_in_sub_document:
+                return "tried to close array while in subdocument";
+            case error_code::k_cannot_close_document_in_sub_array:
+                return "tried to close document while in a subarray";
+            case error_code::k_cannot_perform_array_operation_on_document:
+                return "tried to operate on document, but this is an array";
+            case error_code::k_cannot_perform_document_operation_on_array:
+                return "tried to operate on array, but this is a document";
+#define BSONCXX_ENUM(name, value)                  \
+    case error_code::k_need_element_type_k_##name: \
+        return {"expected element type k_" #name};
+#include <bsoncxx/enums/type.hpp>
+#undef BSONCXX_ENUM
+            case error_code::k_need_key:
+                return "expected a key but found none";
+            case error_code::k_no_array_to_close:
+                return "tried to close array, but no array was opened";
+            case error_code::k_no_document_to_close:
+                return "tried to close document, but no document was opened";
+            case error_code::k_unmatched_key_in_builder:
+                return "can't convert builder to a valid view: unmatched key";
+            case error_code::k_unset_element:
+                return "unset document::element";
+        }
+    }
+};
+
+///
+/// An error category for libbson errors in the BSON_ERROR_JSON domain.
+///
+
+class libbson_json_error final : public std::error_category {
```

Comment:
@acmorrow  and @tessavitabile this has been updated to use separate error categories for libbson error domains, do we like it?